<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>scDHA package manual</title>
<style type="text/css">
/**
 * Prism.s theme ported from highlight.js's xcode style
 */
pre code {
  padding: 1em;
}
.token.comment {
  color: #007400;
}
.token.punctuation {
  color: #999;
}
.token.tag,
.token.selector {
  color: #aa0d91;
}
.token.boolean,
.token.number,
.token.constant,
.token.symbol {
  color: #1c00cf;
}
.token.property,
.token.attr-name,
.token.string,
.token.char,
.token.builtin {
  color: #c41a16;
}
.token.inserted {
  background-color: #ccffd8;
}
.token.deleted {
  background-color: #ffebe9;
}
.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
  color: #9a6e3a;
}
.token.atrule,
.token.attr-value,
.token.keyword {
  color: #836c28;
}
.token.function,
.token.class-name {
  color: #DD4A68;
}
.token.regex,
.token.important,
.token.variable {
  color: #5c2699;
}
.token.important,
.token.bold {
  font-weight: bold;
}
.token.italic {
  font-style: italic;
}
</style>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  box-sizing: border-box;
}
body, .footnotes, code { font-size: .9em; }
li li { font-size: .95em; }
*, *:before, *:after {
  box-sizing: inherit;
}
pre, img { max-width: 100%; }
pre, pre:hover {
  white-space: pre-wrap;
  word-break: break-all;
}
pre code {
  display: block;
  overflow-x: auto;
}
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre) > code, code[class] { background-color: #F8F8F8; }
code.language-undefined, pre > code:not([class]) {
  background-color: inherit;
  border: 1px solid #eee;
}
table {
  margin: auto;
  border-top: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color: #666;
  margin: 0;
  padding-left: 1em;
  border-left: 0.5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC .numbered li { list-style: none; }
#TOC .numbered { padding-left: 0; }
#TOC .numbered ul { padding-left: 1em; }
table, .body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.footnote-ref a::before { content: "["; }
.footnote-ref a::after { content: "]"; }
section.footnotes::before {
  content: "";
  display: block;
  max-width: 20em;
}

@media print {
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  tr, img { page-break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  pre { white-space: pre; }
}
</style>
</head>
<body>
<div class="include-before">
</div>
<div class="frontmatter">
<div class="title"><h1>scDHA package manual</h1></div>
<div class="author"><h2>Duc Tran</h2></div>
<div class="date"><h3>2024-04-02</h3></div>
</div>
<div class="body">
<h1 id="introduction">Introduction</h1>
<p>Advances in microfluidics and sequencing technologies have allowed us to monitor biological systems at single-cell resolution. This comprehensive decomposition of complex tissues holds enormous potential in developmental biology and clinical research. However, the ever-increasing number of cells, technical noise, and high dropout rate pose significant computational challenges in scRNA-seq analysis.</p>
<p>To address this problem, we introduce single-cell Decomposition using Hierarchical Autoencoder (scDHA), which can efficiently detach noise from informative biological signals. The scDHA pipeline consists of two core modules. The first module is a non-negative kernel autoencoder that provides a non-negative, part-based representation of the data. Based on the weight distribution of the encoder, scDHA removes genes or components that have insignificant contributions to the representation. The second module is a Stacked Bayesian Self-learning Network that is built upon the Variational Autoencoder to project the data onto a low dimensional space. Using this informative and compact representation, many analyses can be performed with high accuracy and tractable time complexity (mostly linear or lower complexity).</p>
<p>In one joint framework, the scDHA software package conducts cell segregation through unsupervised learning, dimension reduction and visualization, cell classification, and time-trajectory inference.</p>
<h1 id="installation">Installation</h1>
<h3 id="install-scdha-package-and-other-requirements">Install scDHA package and other requirements</h3>
<p>scDHA can be installed fron GitHub or CRAN using below instruction. scDHA depends on the <code>torch</code> package to build and train the autoencoders. When scDHA package is loaded, it will check for the availability of <code>C++ libtorch</code>. <code>torch</code> package can be used to install <code>C++ libtorch</code>, which is necessary for neural network computation.</p>
<pre><code class="language-r">#Install devtools: 
utils::install.packages('devtools')

#Install the package from GitHub:
devtools::install_github('duct317/scDHA')
#With manual and vignette: 
devtools::install_github('duct317/scDHA', build_manual = T, build_vignettes = T)
#Or from CRAN:
install.packages(&quot;scDHA&quot;)

#When the package is loaded, it will check for C++ libtorch 
library(scDHA)
#libtorch can be installed using:
torch::install_torch()
</code></pre>
<h3 id="install-other-necessary-packages">Install other necessary packages</h3>
<pre><code class="language-r">if (!requireNamespace(&quot;mclust&quot;, quietly = TRUE)) install.packages(&quot;mclust&quot;)
</code></pre>
<h1 id="analysis-on-goolam-dataset">Analysis on Goolam dataset</h1>
<pre><code class="language-r">library(scDHA)
#Load example data (Goolam dataset)
data(&quot;Goolam&quot;)

#Get data matrix and label
data &lt;- t(Goolam$data); label &lt;- as.character(Goolam$label)

#Log transform the data 
data &lt;- log2(data + 1)
</code></pre>
<h3 id="clustering">Clustering</h3>
<pre><code class="language-r">#Generate clustering result, the input matrix has rows as samples and columns as genes
result &lt;- scDHA(data, seed = 1)

#The clustering result can be found here 
cluster &lt;- result$cluster

#Calculate adjusted Rand Index using mclust package
ari &lt;- round(mclust::adjustedRandIndex(cluster,label), 2)
print(paste0(&quot;ARI = &quot;, ari))
</code></pre>
<h3 id="visualization">Visualization</h3>
<pre><code class="language-r">#Generate 2D representation, the input is the output from scDHA function
result &lt;- scDHA.vis(result, seed = 1)

#Plot the representation of the dataset, different colors represent different cell types
plot(result$pred, col=factor(label), xlab = &quot;scDHA1&quot;, ylab = &quot;scDHA2&quot;)
</code></pre>
<h3 id="time-trajectory-inference">Time trajectory inference</h3>
<pre><code class="language-r">#Cell stage order in Goolam dataset
cell.stages &lt;- c(&quot;2cell&quot;, &quot;4cell&quot;, &quot;8cell&quot;, &quot;16cell&quot;, &quot;blast&quot;)

#Generate pseudo-time for each cell, the input is the output from scDHA function
result &lt;- scDHA.pt(result, start.point = 1, seed = 1)

#Calculate R-squared value representing correlation between inferred pseudo-time and cell stage order
r2 &lt;- round(cor(result$pt, as.numeric(factor(label, levels = cell.stages)))^2, digits = 2)

#Plot pseudo-temporal ordering of cells in Goolam dataset
plot(result$pt, factor(label, levels = cell.stages), xlab= &quot;Pseudo Time&quot;, ylab = &quot;Cell Stages&quot;, xaxt=&quot;n&quot;, yaxt=&quot;n&quot;)
axis(2, at=1:5,labels=cell.stages, las=2)
text(x = 1, y = 4.5, labels = paste0(&quot;R2 = &quot;, r2))
</code></pre>
<h3 id="cell-classification">Cell classification</h3>
<pre><code class="language-r">#Split data into training and testing sets
set.seed(1)
idx &lt;- sample.int(nrow(data), size = round(nrow(data)*0.75))

train.x &lt;- data[idx, ]; train.y &lt;- label[idx]
test.x &lt;- data[-idx, ]; test.y &lt;- label[-idx]

#Predict the labels of cells in testing set, the input matrices have rows as samples and columns as genes
prediction &lt;- scDHA.class(train = train.x, train.label = train.y, test = test.x, seed = 1)

#Calculate accuracy of the predictions
acc &lt;- round(sum(test.y == prediction)/length(test.y), 2)
print(paste0(&quot;Accuracy = &quot;, acc))
</code></pre>
</div>
<div class="include-after">
</div>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</body>
</html>
