<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="generator" content="litedown 0.7">
<title>scDHA package manual</title>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  print-color-adjust: exact;
  -webkit-print-color-adjust: exact;
}
body, .abstract, code, .footnotes, footer, #refs, .caption { font-size: .9em; }
li li { font-size: .95em; }
ul:has(li > input[type="checkbox"]) { list-style: none; padding-left: 1em; }
*, :before, :after { box-sizing: border-box; }
a { color: steelblue; }
pre, img { max-width: 100%; }
pre { white-space: pre-wrap; word-break: break-word; }
pre code { display: block; padding: 1em; overflow-x: auto; }
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre, th) > code, code[class], div > .caption { background: #f8f8f8; }
pre > code:is(:not([class]), .language-plain, .language-none, .plain), .box, .figure, .table { background: inherit; border: 1px solid #eee; }
pre > code {
  &.message { border-color: #9eeaf9; }
  &.warning { background: #fff3cd; border-color: #fff3cd; }
  &.error { background: #f8d7da; border-color: #f8d7da; }
}
.fenced-chunk { border-left: 1px solid #666; }
.code-fence {
  opacity: .4;
  border: 1px dashed #666;
  border-left: 2px solid;
  &:hover { opacity: inherit; }
}
.box, .figure, .table, table { margin: 1em auto; }
div > .caption { padding: 1px 1em; }
.figure { p:has(img, svg), pre:has(svg) { text-align: center; } }
.flex-col { display: flex; justify-content: space-between; }
table {
  &:only-child:not(.table > *) { margin: auto; }
  th, td { padding: 5px; font-variant-numeric: tabular-nums; }
  thead, tfoot, tr:nth-child(even) { background: whitesmoke; }
  thead th { border-bottom: 1px solid #ddd; }
  &:not(.datatable-table) {
    border-top: 1px solid #666;
    border-bottom: 1px solid #666;
  }
}
blockquote {
  color: #666;
  margin: 0;
  padding: 1px 1em;
  border-left: .5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC {
  a { text-decoration: none; }
  ul { list-style: none; padding-left: 1em; }
  & > ul { padding: 0; }
  ul ul { border-left: 1px solid lightsteelblue; }
}
.body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.main-number::after { content: "."; }
span[class^="ref-number-"] { font-weight: bold; }
.ref-number-fig::after, .ref-number-tab::after { content: ":"; }
.cross-ref-chp::before { content: "Chapter "; }
.cross-ref-sec::before { content: "Section "; }
.cross-ref-fig::before, .ref-number-fig::before { content: "Figure "; }
.cross-ref-tab::before, .ref-number-tab::before { content: "Table "; }
.cross-ref-eqn::before, .MathJax_ref:has(mjx-mtext > mjx-c + mjx-c)::before { content: "Equation "; }
.abstract, #refs {
  &::before { display: block; margin: 1em auto; font-weight: bold; }
}
.abstract::before { content: "Abstract"; text-align: center; }
#refs::before { content: "Bibliography"; font-size: 1.5em; }
.ref-paren-open::before { content: "("; }
.ref-paren-close::after { content: ")"; }
.ref-semicolon::after { content: "; "; }
.ref-and::after { content: " and "; }
.ref-et-al::after { content: " et al."; font-style: italic; }
.footnote-ref a {
  &::before { content: "["; }
  &::after { content: "]"; }
}
section.footnotes {
  margin-top: 2em;
  &::before { content: ""; display: block; max-width: 20em; }
}
.fade {
  background: repeating-linear-gradient(135deg, white, white 30px, #ddd 32px, #ddd 32px);
  opacity: 0.6;
}

@media print {
  body { max-width: 100%; }
  tr, img { break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  body:not(.pagesjs) pre:has(.line-numbers):not(:hover) { white-space: pre; }
}
</style>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@xiee/utils@1.14.14/css/prism-xcode.min.css">
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</head>
<body>
<div class="frontmatter">
<div class="title"><h1>scDHA package manual</h1></div>
<div class="author"><h2>Duc Tran</h2></div>
<div class="date"><h3>2025-09-23</h3></div>
</div>
<div class="body">
<h1 id="chp:introduction">Introduction</h1>
<p>Advances in microfluidics and sequencing technologies have allowed us to monitor biological systems at single-cell resolution. This comprehensive decomposition of complex tissues holds enormous potential in developmental biology and clinical research. However, the ever-increasing number of cells, technical noise, and high dropout rate pose significant computational challenges in scRNA-seq analysis.</p>
<p>To address this problem, we introduce single-cell Decomposition using Hierarchical Autoencoder (scDHA), which can efficiently detach noise from informative biological signals. The scDHA pipeline consists of two core modules. The first module is a non-negative kernel autoencoder that provides a non-negative, part-based representation of the data. Based on the weight distribution of the encoder, scDHA removes genes or components that have insignificant contributions to the representation. The second module is a Stacked Bayesian Self-learning Network that is built upon the Variational Autoencoder to project the data onto a low dimensional space. Using this informative and compact representation, many analyses can be performed with high accuracy and tractable time complexity (mostly linear or lower complexity).</p>
<p>In one joint framework, the scDHA software package conducts cell segregation through unsupervised learning, dimension reduction and visualization, cell classification, and time-trajectory inference.</p>
<h1 id="chp:installation">Installation</h1>
<h3 id="sec:install-scdha-package-and-other-requirements">Install scDHA package and other requirements</h3>
<p>scDHA can be installed fron GitHub or CRAN using below instruction. scDHA depends on the <code>torch</code> package to build and train the autoencoders. When scDHA package is loaded, it will check for the availability of <code>C++ libtorch</code>. <code>torch</code> package can be used to install <code>C++ libtorch</code>, which is necessary for neural network computation.</p>
<pre><code class="language-r">#Install devtools: 
utils::install.packages('devtools')

#Install the package from GitHub:
devtools::install_github('duct317/scDHA')
#With manual and vignette: 
devtools::install_github('duct317/scDHA', build_manual = T, build_vignettes = T)
#Or from CRAN:
install.packages(&quot;scDHA&quot;)

#When the package is loaded, it will check for C++ libtorch 
library(scDHA)
#libtorch can be installed using:
torch::install_torch()
</code></pre>
<h3 id="sec:install-other-necessary-packages">Install other necessary packages</h3>
<pre><code class="language-r">if (!requireNamespace(&quot;mclust&quot;, quietly = TRUE)) install.packages(&quot;mclust&quot;)
</code></pre>
<h1 id="chp:analysis-on-goolam-dataset">Analysis on Goolam dataset</h1>
<pre><code class="language-r">library(scDHA)
#Load example data (Goolam dataset)
data(&quot;Goolam&quot;)

#Get data matrix and label
data &lt;- t(Goolam$data); label &lt;- as.character(Goolam$label)

#Log transform the data 
data &lt;- log2(data + 1)
</code></pre>
<h3 id="sec:clustering">Clustering</h3>
<pre><code class="language-r">#Generate clustering result, the input matrix has rows as samples and columns as genes
result &lt;- scDHA(data, seed = 1)

#The clustering result can be found here 
cluster &lt;- result$cluster

#Calculate adjusted Rand Index using mclust package
ari &lt;- round(mclust::adjustedRandIndex(cluster,label), 2)
print(paste0(&quot;ARI = &quot;, ari))
</code></pre>
<h3 id="sec:visualization">Visualization</h3>
<pre><code class="language-r">#Generate 2D representation, the input is the output from scDHA function
result &lt;- scDHA.vis(result, seed = 1)

#Plot the representation of the dataset, different colors represent different cell types
plot(result$pred, col=factor(label), xlab = &quot;scDHA1&quot;, ylab = &quot;scDHA2&quot;)
</code></pre>
<h3 id="sec:time-trajectory-inference">Time trajectory inference</h3>
<pre><code class="language-r">#Cell stage order in Goolam dataset
cell.stages &lt;- c(&quot;2cell&quot;, &quot;4cell&quot;, &quot;8cell&quot;, &quot;16cell&quot;, &quot;blast&quot;)

#Generate pseudo-time for each cell, the input is the output from scDHA function
result &lt;- scDHA.pt(result, start.point = 1, seed = 1)

#Calculate R-squared value representing correlation between inferred pseudo-time and cell stage order
r2 &lt;- round(cor(result$pt, as.numeric(factor(label, levels = cell.stages)))^2, digits = 2)

#Plot pseudo-temporal ordering of cells in Goolam dataset
plot(result$pt, factor(label, levels = cell.stages), xlab= &quot;Pseudo Time&quot;, ylab = &quot;Cell Stages&quot;, xaxt=&quot;n&quot;, yaxt=&quot;n&quot;)
axis(2, at=1:5,labels=cell.stages, las=2)
text(x = 1, y = 4.5, labels = paste0(&quot;R2 = &quot;, r2))
</code></pre>
<h3 id="sec:cell-classification">Cell classification</h3>
<pre><code class="language-r">#Split data into training and testing sets
set.seed(1)
idx &lt;- sample.int(nrow(data), size = round(nrow(data)*0.75))

train.x &lt;- data[idx, ]; train.y &lt;- label[idx]
test.x &lt;- data[-idx, ]; test.y &lt;- label[-idx]

#Predict the labels of cells in testing set, the input matrices have rows as samples and columns as genes
prediction &lt;- scDHA.class(train = train.x, train.label = train.y, test = test.x, seed = 1)

#Calculate accuracy of the predictions
acc &lt;- round(sum(test.y == prediction)/length(test.y), 2)
print(paste0(&quot;Accuracy = &quot;, acc))
</code></pre>
</div>
</body>
</html>
